tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
+
+  auto loc = op->getLoc();
+  auto rank = value_ty.getRank();
+  // Get int vector [0, 1, ..., rank-1]
+  SmallVector<Value, 4> dim_sizes;
+  for (size_t d = 0; d < rank; ++d) {
+    dim_sizes.emplace_back(rewriter.create<tensor::DimOp>(loc, value, d));
+  }
+  return dim_sizes;
+}
+
+Value unsqueezeTensorDim(PatternRewriter& rewriter, Operation* op, Value tensor,
+                         int64_t dim) {
+  // Returns a new tensor with dims of size 1 inserted at the specified
+  // position.
+  //
+  // The position indices (must be high to low dimension number of the returned
+  // tensor) are specified with unsqzDims. Indices must be in-order, and in
+  // range of tensor rank. Thus, unsqueeze a rank 1 tensor with {0, 2}, {0, 1,
+  // 3}, {0, 1, 2} are all valid dimension sets, but {0, 3}, {2} are not.
+  auto dim_sizes = getDimSizesOfTensor(rewriter, op, tensor);
+  auto loc = op->getLoc();
+  auto one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+  auto rank = dim_sizes.size();
+
+  auto rankTy = tensor.getType().dyn_cast<RankedTensorType>();
+  auto oldShape = rankTy.getShape();
+
+  SmallVector<Value, 4> newDimSizes;
+  std::vector<int64_t> newShape;
+  newDimSizes.reserve(rank + 1);
+  newShape.reserve(rank + 1);
+  for (size_t k = 0, i = 0; k < rank + 1; ++k) {
+    if (dim == k) {
+      newDimSizes.push_back(one);
+      newShape.push_back(1);
+    } else {
+      newDimSizes.push_back(dim_sizes[i]);
+      newShape.push_back(oldShape[i]);
+      i++;
+    }
+  }
+
+  auto outTy = RankedTensorType::get(newShape, rankTy.getElementType());
+  auto mhloShape = rewriter.create<tensor::FromElementsOp>(loc, newDimSizes);
+  return rewriter.create<mhlo::DynamicReshapeOp>(loc, outTy, tensor, mhloShape)
+      .getResult();
+}
+
+Value squeezeTensorDim(PatternRewriter& rewriter, Operation* op, Value tensor,
+                       int64_t dim) {
+  auto dim_sizes = getDimSizesOfTensor(rewriter, op, tensor);
+  auto rank = dim_sizes.size();
+
+  auto rankTy = tensor.getType().dyn_cast<RankedTensorType>();
+  auto oldShape = rankTy.getShape();
+
+  SmallVector<Value, 4> newDimSizes;
+  std::vector<int64_t> newShape;
+  newDimSizes.reserve(rank - 1);
+  newShape.reserve(rank - 1);
+  for (size_t k = 0; k < rank; ++k) {
+    if (dim == k) continue;
+    newDimSizes.push_back(dim_sizes[k]);
+    newShape.push_back(oldShape[k]);
+  }
+
+  auto outTy = RankedTensorType::get(newShape, rankTy.getElementType());
+  auto loc = op->getLoc();
+  auto mhloShape = rewriter.create<tensor::FromElementsOp>(loc, newDimSizes);
+  return rewriter.create<mhlo::DynamicReshapeOp>(loc, outTy, tensor, mhloShape)
+      .getResult();
+}
+
+// Converts DotGeneralOp if it represents vec * vec, vec * mat, mat * vec.
+// Will first unsqueeze vector to matirx, then convert DoGeneralOp to
+// matrix multiply; After that drop(unsqueeze) the extra-dimensions from
+// the output of DoGeneralOp.
+struct DotGeneralConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
```

Comment:
Shall we have a .mlir test for this conversion on compiler side?

Reply:
+1